seed/seedwriter/writer.go

```diff
@@ -921,3 +941,58 @@ func (w *Writer) WriteMeta() error {
 
 	return w.tree.writeMeta(snapsFromModel, extraSnaps)
 }
+
+// query accessors
+
+func (w *Writer) checkSnapsAccessor() error {
+	if w.expectedStep < seedSnapsStep {
+		return fmt.Errorf("internal error: seedwriter.Writer cannot query seed snaps before Downloaded signaled complete")
+	}
+	return nil
+}
+
+// BootSnaps returns the seed snaps involved in the boot process.
+// It can be invoked only after Downloaded returns complete ==
+// true. It returns an error for classic models as for those no snaps
+// participate in boot before user space.
+func (w *Writer) BootSnaps() ([]*SeedSnap, error) {
+	if err := w.checkSnapsAccessor(); err != nil {
+		return nil, err
+	}
+	if w.model.Classic() {
+		return nil, fmt.Errorf("no snaps participating in boot on classic")
+	}
+	var bootSnaps []*SeedSnap
+	for _, sn := range w.snapsFromModel {
+		bootSnaps = append(bootSnaps, sn)
+		if sn.Info.GetType() == snap.TypeGadget {
+			break
+
+		}
+	}
+	return bootSnaps, nil
+}
+
+// UnassertedSnaps returns references for all unasserted snaps in the seed.
+// It can be invoked only after Downloaded returns complete ==
+// true.
+func (w *Writer) UnassertedSnaps() ([]naming.SnapRef, error) {
```

Comment:
Small wondering (without looking at the rest of the code) if we could have consistency here wrt returned types in `BoostSnaps()` and `UnassertedSnaps` (i.e. []*SeedSnap vs []naming.SnapRef). Is it possible to infer one from the other and use more generic one for both functions?

Reply:
it's an encapsulation decision, one return type says the code expects you to need and care about the details,  the other expects that knowing the "names" is enough